pkg/server/plugin/keymanager/awskms/awskms.go

```diff
@@ -0,0 +1,533 @@
+package awskms
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/andres-erbsen/clock"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/kms"
+	"github.com/aws/aws-sdk-go-v2/service/kms/types"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/hcl"
+	"github.com/spiffe/spire/pkg/common/catalog"
+	"github.com/spiffe/spire/pkg/server/plugin/keymanager"
+	"github.com/spiffe/spire/proto/spire/common/plugin"
+	"github.com/zeebo/errs"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	pluginName       = "awskms"
+	aliasPrefix      = "alias/"
+	defaultKeyPrefix = "SPIRE_SERVER_KEY/"
+
+	keyIDTag = "key_id"
+	aliasTag = "alias"
+)
+
+var (
+	kmsErr = errs.Class(pluginName)
+)
+
+func BuiltIn() catalog.Plugin {
+	return builtin(New())
+}
+
+func builtin(p *Plugin) catalog.Plugin {
+	return catalog.MakePlugin(pluginName, keymanager.PluginServer(p))
+}
+
+type keyEntry struct {
+	KMSKeyID  string
+	Alias     string
+	PublicKey *keymanager.PublicKey
+}
+
+// Plugin is the main representation of this keymanager plugin
+type Plugin struct {
+	keymanager.UnsafeKeyManagerServer
+	log            hclog.Logger
+	mu             sync.RWMutex
+	entries        map[string]keyEntry
+	kmsClient      kmsClient
+	keyPrefix      string
+	scheduleDelete chan string
+	clock          clock.Clock
+
+	hooks struct {
+		newClient func(ctx context.Context, config *Config) (kmsClient, error)
+	}
+}
+
+// Config provides configuration context for the plugin
+type Config struct {
+	AccessKeyID     string `hcl:"access_key_id" json:"access_key_id"`
+	SecretAccessKey string `hcl:"secret_access_key" json:"secret_access_key"`
+	Region          string `hcl:"region" json:"region"`
+	KeyPrefix       string `hcl:"key_prefix" json:"key_prefix"`
+}
+
+// New returns an instantiated plugin
+func New() *Plugin {
+	return newPlugin(newKMSClient)
+}
+
+func newPlugin(newClient func(ctx context.Context, config *Config) (kmsClient, error)) *Plugin {
+	p := &Plugin{}
+	p.hooks.newClient = newClient
+	p.entries = make(map[string]keyEntry)
+	p.scheduleDelete = make(chan string, 120)
+	p.clock = clock.New()
+	return p
+}
+
+// SetLogger sets a logger
+func (p *Plugin) SetLogger(log hclog.Logger) {
+	p.log = log
+}
+
+// Configure sets up the plugin
+func (p *Plugin) Configure(ctx context.Context, req *plugin.ConfigureRequest) (*plugin.ConfigureResponse, error) {
+	config, err := p.validateConfig(req.Configuration)
+	if err != nil {
+		return nil, err
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.keyPrefix = config.KeyPrefix
+	p.kmsClient, err = p.hooks.newClient(ctx, config)
+	if err != nil {
+		return nil, kmsErr.New("failed to create KMS client: %v", err)
+	}
+
+	p.log.Debug("Fetching key aliases from KMS")
+	var nextMarker *string
+	for {
+		nextMarker, err = p.fetchAliasesPage(ctx, nextMarker)
+		if err != nil {
+			return nil, err
+		}
+		if nextMarker == nil {
+			break
+		}
+	}
+
+	go p.scheduleDeleteTask(ctx)
```

Comment:
If you use this context for the task, I believe the context will be canceled as soon as the RPC has completed and the delete task will stop.... we should use a `context.WithCancel(context.Background())` for this, and hold onto the cancellation callback so you can cancel it on a re-configure.

Reply:
That's exactly what is happening. Thanks for that suggestion!